### PR TITLE
gdal: add java bindings

### PR DESCRIPTION
**Pull request by rollf**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by imincik** [comment]
Thank you for contribution. This PR is [failing on Darwin](https://github.com/NixOS/nixpkgs/pull/280279/checks?check_run_id=20395778616) now.

File: pkgs/development/libraries/gdal/tests.nix
Comment (imincik) [inline pkgs/development/libraries/gdal/tests.nix:27]: Would you please move java binding tests to the very bottom of the test script ?

File: pkgs/development/libraries/gdal/tests.nix
Comment (imincik) [inline pkgs/development/libraries/gdal/tests.nix:7]: If possible, please avoid reformatting of whole file.

Comment (imincik) [inline pkgs/development/libraries/gdal/tests.nix:7]: Please avoid reformatting of whole file if not necessary.

**Comment by imincik** [comment]
Maybe we should also consider building with java as optional feature using `withJava ? false` derivation parameter.

**Comment by imincik** [comment]
Result of `nixpkgs-review pr 280279` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.worldengine</li>
    <li>python310Packages.worldengine.dist</li>
    <li>python311Packages.worldengine</li>
    <li>python311Packages.worldengine.dist</li>
    <li>t-rex</li>
    <li>worldengine-cli</li>
    <li>worldengine-cli.dist</li>
  </ul>
</details>
<details>
  <summary>24 packages failed to build:</summary>
  <ul>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>python310Packages.bsuite</li>
    <li>python310Packages.bsuite.dist</li>
    <li>python310Packages.cartopy</li>
    <li>python310Packages.cartopy.dist</li>
    <li>python310Packages.django-bootstrap4</li>
    <li>python310Packages.django-bootstrap4.dist</li>
    <li>python310Packages.fiona</li>
    <li>python310Packages.fiona.dist</li>
    <li>python310Packages.folium</li>
    <li>python310Packages.folium.dist</li>
    <li>python310Packages.gdal</li>
    <li>python310Packages.geopandas</li>
    <li>python310Packages.geopandas.dist</li>
    <li>python310Packages.osmnx</li>
    <li>python310Packages.osmnx.dist</li>
    <li>python310Packages.plotnine</li>
    <li>python310Packages.plotnine.dist</li>
    <li>python310Packages.rasterio</li>
    <li>python310Packages.rasterio.dist</li>
    <li>python310Packages.wktutils</li>
    <li>python310Packages.wktutils.dist</li>
    <li>pytrainer</li>
    <li>pytrainer.dist</li>
  </ul>
</details>
<details>
  <summary>79 packages built:</summary>
  <ul>
    <li>cloudcompare</li>
    <li>entwine</li>
    <li>gdal (python311Packages.gdal)</li>
    <li>gdalMinimal</li>
    <li>gmt</li>
    <li>gplates</li>
    <li>grass</li>
    <li>haskellPackages.hgdal</li>
    <li>haskellPackages.hgdal.doc</li>
    <li>mapcache</li>
    <li>mapnik</li>
    <li>mapproxy</li>
    <li>mapproxy.dist</li>
    <li>mapserver</li>
    <li>merkaartor</li>
    <li>mysql-workbench</li>
    <li>octavePackages.mapping</li>
    <li>openorienteering-mapper</li>
    <li>paraview</li>
    <li>pdal</li>
    <li>perl536Packages.Tirex</li>
    <li>perl536Packages.Tirex.devdoc</li>
    <li>perl538Packages.Tirex</li>
    <li>perl538Packages.Tirex.devdoc</li>
    <li>postgresql12JitPackages.postgis</li>
    <li>postgresql12JitPackages.postgis.doc</li>
    <li>postgresql12Packages.postgis</li>
    <li>postgresql12Packages.postgis.doc</li>
    <li>postgresql13JitPackages.postgis</li>
    <li>postgresql13JitPackages.postgis.doc</li>
    <li>postgresql13Packages.postgis</li>
    <li>postgresql13Packages.postgis.doc</li>
    <li>postgresql14JitPackages.postgis</li>
    <li>postgresql14JitPackages.postgis.doc</li>
    <li>postgresql14Packages.postgis</li>
    <li>postgresql14Packages.postgis.doc</li>
    <li>postgresqlJitPackages.postgis (postgresql15JitPackages.postgis)</li>
    <li>postgresqlJitPackages.postgis.doc (postgresql15JitPackages.postgis.doc)</li>
    <li>postgresqlPackages.postgis (postgresql15Packages.postgis)</li>
    <li>postgresqlPackages.postgis.doc (postgresql15Packages.postgis.doc)</li>
    <li>postgresql16JitPackages.postgis</li>
    <li>postgresql16JitPackages.postgis.doc</li>
    <li>postgresql16Packages.postgis</li>
    <li>postgresql16Packages.postgis.doc</li>
    <li>python310Packages.pygmt</li>
    <li>python310Packages.pygmt.dist</li>
    <li>python310Packages.python-mapnik</li>
    <li>python310Packages.python-mapnik.dist</li>
    <li>python311Packages.bsuite</li>
    <li>python311Packages.bsuite.dist</li>
    <li>python311Packages.cartopy</li>
    <li>python311Packages.cartopy.dist</li>
    <li>python311Packages.django-bootstrap4</li>
    <li>python311Packages.django-bootstrap4.dist</li>
    <li>python311Packages.fiona</li>
    <li>python311Packages.fiona.dist</li>
    <li>python311Packages.folium</li>
    <li>python311Packages.folium.dist</li>
    <li>python311Packages.geopandas</li>
    <li>python311Packages.geopandas.dist</li>
    <li>python311Packages.osmnx</li>
    <li>python311Packages.osmnx.dist</li>
    <li>python311Packages.plotnine</li>
    <li>python311Packages.plotnine.dist</li>
    <li>python311Packages.pygmt</li>
    <li>python311Packages.pygmt.dist</li>
    <li>python311Packages.python-mapnik</li>
    <li>python311Packages.python-mapnik.dist</li>
    <li>python311Packages.rasterio</li>
    <li>python311Packages.rasterio.dist</li>
    <li>python311Packages.wktutils</li>
    <li>python311Packages.wktutils.dist</li>
    <li>qgis</li>
    <li>qgis-ltr</li>
    <li>qmapshack</li>
    <li>saga</li>
    <li>sumo</li>
    <li>udig</li>
    <li>vpv</li>
  </ul>
</details>

File: pkgs/development/libraries/gdal/tests.nix
Comment (rollf) [inline pkgs/development/libraries/gdal/tests.nix:25]: This example is based on https://gdal.org/api/java/index.html#how-to-use-the-bindings.

**Comment by rollf** [comment]
> Maybe we should also consider building with java as optional feature using `withJava ? false` derivation parameter.

@imincik Thanks for your preliminary review. I will implement this.

File: pkgs/development/libraries/gdal/default.nix
Comment (rollf) [inline pkgs/development/libraries/gdal/default.nix:159]: Sorry, this gets auto-intended all the time here.

**Comment by r-burns** [comment]
> Maybe we should also consider building with java as optional feature using `withJava ? false` derivation parameter.

Alternatively, if it's possible to put the java bindings in a separate output, we could always build the java bindings and not worry about closure bloat in environments containing gdal with+without java support, as they'd both use the same underlying C++ library. But if it's difficult to split out a `-java` output then I think `withJava` is a good approach.

**Comment by rollf** [comment]
The GDAL teams defaults to building it Java (if `ant` & `jdk` are available) and we already have a `withX` approach for other optional features so I tend to go with `withJava` to keep things consistent.

That being said, I'm not the maintainer.

**Comment by rollf** [comment]
Look like this doesn't build on darwin. From the ofBorg [log](https://github.com/NixOS/nixpkgs/pull/280279/checks?check_run_id=20428125810):


```
       > Center      (  400.0,  300.0)
       > Band 1 Block=800x5 Type=UInt16, ColorInterp=Gray
       >   NoData Value=255
       > Native library load failed.
       > java.lang.UnsatisfiedLinkError: no gdalalljni in java.library.path: /nix/store/iyysy6nv6w55cy2lkignza88hdascpa9-gdal-3.8.2/lib/
       > Exception in thread "main" java.lang.UnsatisfiedLinkError: 'void org.gdal.gdal.gdalJNI.AllRegister()'
       >    at org.gdal.gdal.gdalJNI.AllRegister(Native Method)
       >    at org.gdal.gdal.gdal.AllRegister(gdal.java:712)
       >       at Main.main(main.java:4)
       For full logs, run 'nix log /nix/store/43ryfz175vlkmg60i43h79nraifkxml6-gdal-tests.drv'.
```

Seems like ti `.so` file could not be built.

There are several `errors` in the [full log](https://logs.ofborg.org/?key=nixos/nixpkgs.280279&attempt_id=3729e557-563b-4c57-903c-4cdd7329c7b9), like:

```
/tmp/nix-build-gdal-3.8.2.drv-1/source/build/swig/java/org_patched/org/gdal/ogr/ogr.java:17: error: invalid uri: "#Open(java.lang.String, int)"
OR
/tmp/nix-build-gdal-3.8.2.drv-1/source/build/swig/java/org_patched/org/gdal/gdal/Band.java:514: error: invalid uri: "#ReadBlock_Direct(int, int, java.nio.ByteBuffer)"
```

I have no way to inspect this on a Mac.

**Comment by rollf** [comment]
Ah, no that I think about this. I guess it worked just fine because of the log line 

```
Installing: /nix/store/iyysy6nv6w55cy2lkignza88hdascpa9-gdal-3.8.2/lib/jni/libgdalalljni.dylib
```

but the test fails. I don't have experience with Mac.

**Comment by rollf** [comment]
Okay, I will see what I can do. @imincik FYI, I originally worked on this [PR](https://github.com/NixOS/nixpkgs/pull/281739). It is not prepared yet, I just opened a draft PR to be able to point to it from here. Working on this caused me to add the Java bindings to `gdal` (hence this PR here)

**Comment by rollf** [comment]
@imincik I managed to fix the Darwin issue. The problem was that I symlinked a `*.so` file although on Darwin, these are called `*.dylib`. I now use a simple shell expansion. All checks of this PR pass now (at least for [dd8dcd0](https://github.com/NixOS/nixpkgs/pull/280279/commits/dd8dcd04d5880e95011be470b0e1069ab1dd1c4f); currently [35353ca](https://github.com/NixOS/nixpkgs/pull/280279/commits/35353caa69a000891b0cc9f6887f48d4802dc6cc) gets evaluated but I'm confident it will pass, too). Do you mind having another look? Once this is merged, I'd like to polish https://github.com/NixOS/nixpkgs/pull/281739.

**Comment by rollf** [comment]
Phew ... now that I found some debug leftovers, should I squash this whole PR?

File: pkgs/development/libraries/gdal/default.nix
Comment (imincik) [inline pkgs/development/libraries/gdal/default.nix:17]: I compared closure sizes of gdal built with Java and without and the one built with Java is only little bit bigger. 

```
nix-build -E 'with import ./. {}; gdal.override { useJava = false; }'

nix path-info -rsSh /nix/store/wr943gniijj42v46ym1n31pcqz5mixky-gdal-3.8.3

/nix/store/wr943gniijj42v46ym1n31pcqz5mixky-gdal-3.8.3                        	  43.2M	   1.3G
```
```
nix-build -E 'with import ./. {}; gdal.override { useJava = true; }'

nix path-info -rsSh /nix/store/03l8izqrs7ni4dbizv4cbainl2p3nk0k-gdal-3.8.3

/nix/store/03l8izqrs7ni4dbizv4cbainl2p3nk0k-gdal-3.8.3                        	  45.2M	   1.3G
```

But even with such a small difference, I suggest building without Java by default due to larger size of build dependencies (`ant` and `jdk`).

```suggestion
, useJava ? false
```

File: pkgs/development/libraries/gdal/tests.nix
Comment (imincik) [inline pkgs/development/libraries/gdal/tests.nix:43]: I guess this test will fail if we build without Java by default.

**Review by imincik — CHANGES_REQUESTED** [review]
Thank you very much for your effort @rollf .

I requested some changes. My biggest concern is whether we should build GDAL with java by default or not.

Also, please remove trailing dot (`.`) from the commit message and don't use capital letters in change description if not needed (suggested commit message should be something like `gdal: add java bindings`).

@thoughtpolice , as the main author of `gdalMinimal` would you mind to review this PR ?

**Comment by imincik** [comment]
@rollf , please remove trailing dot (`.`) from the commit message and don't use capital letters in change description if not needed (suggested commit message should be something like `gdal: add java bindings`). Thank you.

**Comment by rollf** [comment]
@imincik Thank you for the review. I  addressed your change requests except for `withJava ? false`. I currently wait for the ofBorg results and fix issues if any.. I appreciate the attention paid to the closure size. However, I think I have convincing arguments to build with Java support by default:

* GDAL ships with a lot of possible integraitons; from what I can see, they are mostly build automatically when the deps are found. For example, the nix package definition adds `cfitsio`, `brunsli`, `cryptopp`, `geos`, etc. as dependencies which in turn enables the usage of these libraries. Python bindings are built by default if python is found and `python3` is yet another dependency provided in `gdal`'s package definition. Why treat Java differently? Just because it came late to the party? I feel like `useJava ? (!useMinimalFeatures)` would be a good compromise here.
* The first package merged into `nixpkgs` that uses the GDAL java bindings will have to use `gdal.override { withJava = true; }` which will actually render the size argument obsolute since it will essentially force the existence 3 gdal packages: 
   * "full gdal (with Java, not build by default)",
   * "full gdal (without Java, build by default)", 
   * and "minimal gdal (build by default)".
* Any packages that uses the non-default-gdal-with-Java derivation needs to build gdal (which is actually quite time-consuming.
* The test code I added will fail by default, so
  * either I remove the test code which means I'd contribute non-tested code. Future maintainers will have to make sure the building of the Java bindings works
  * or I simply use `gdal.override { withJava = true; }` which creates a new package solely for the test (see my 2nd argument).

Please consider this carefully. As I mentioned [above](https://github.com/NixOS/nixpkgs/pull/280279#issuecomment-1898133140), I plan to use gdal-with-Java-bindings rather soon.

**Comment by imincik** [comment]
> I addressed your change requests except for withJava ? false

I must say I like your arguments. Personally, I am happy to take it, but I would be happier if more people will say yes.

**Comment by das-g** [comment]
I'm no expert on closure size considerations, but @rollf's arguments convince me. Strongly.

If still in doubt, I'd opt for the `useJava ? (!useMinimalFeatures)` compromise he suggests.

File: pkgs/development/libraries/gdal/default.nix
Comment (rollf) [reply pkgs/development/libraries/gdal/default.nix:155]: A separated these changes into another commit. I used `nixpkgs-fmt`.

**Comment by rollf** [comment]
> I must say I like your arguments. Personally, I am happy to take it, but I would be happier if more people will say yes.

> I'm no expert on closure size considerations, but @rollf's arguments convince me. Strongly.
> 
> If still in doubt, I'd opt for the `useJava ? (!useMinimalFeatures)` compromise he suggests.

Okay, thank you.  I believe I have addressed all other change requests. From my point of view, this PR is done and could be merged.

**Comment by imincik** [comment]
Current default Java version is EOL - https://github.com/NixOS/nixpkgs/issues/280901 . Fix in https://github.com/NixOS/nixpkgs/pull/273811

**Comment by imincik** [comment]
Result of `nixpkgs-review pr 280279` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.worldengine</li>
    <li>python311Packages.worldengine.dist</li>
    <li>python312Packages.fiona</li>
    <li>python312Packages.fiona.dist</li>
    <li>python312Packages.geopandas</li>
    <li>python312Packages.geopandas.dist</li>
    <li>python312Packages.plotnine</li>
    <li>python312Packages.plotnine.dist</li>
    <li>python312Packages.worldengine</li>
    <li>python312Packages.worldengine.dist</li>
    <li>t-rex</li>
    <li>worldengine-cli</li>
    <li>worldengine-cli.dist</li>
  </ul>
</details>
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>cloudcompare</li>
    <li>python311Packages.osmnx</li>
    <li>python311Packages.osmnx.dist</li>
    <li>python311Packages.rasterio</li>
    <li>python311Packages.rasterio.dist</li>
    <li>qgis-ltr</li>
  </ul>
</details>
<details>
  <summary>75 packages built:</summary>
  <ul>
    <li>entwine</li>
    <li>flatcam</li>
    <li>gdal (python311Packages.gdal)</li>
    <li>gdalMinimal</li>
    <li>gmt</li>
    <li>gplates</li>
    <li>grass</li>
    <li>haskellPackages.hgdal</li>
    <li>haskellPackages.hgdal.doc</li>
    <li>mapcache</li>
    <li>mapnik</li>
    <li>mapproxy</li>
    <li>mapproxy.dist</li>
    <li>mapserver</li>
    <li>merkaartor</li>
    <li>mysql-workbench</li>
    <li>octavePackages.mapping</li>
    <li>openorienteering-mapper</li>
    <li>paraview</li>
    <li>pdal</li>
    <li>perl536Packages.Tirex</li>
    <li>perl536Packages.Tirex.devdoc</li>
    <li>perl538Packages.Tirex</li>
    <li>perl538Packages.Tirex.devdoc</li>
    <li>postgresql12JitPackages.postgis</li>
    <li>postgresql12JitPackages.postgis.doc</li>
    <li>postgresql12Packages.postgis</li>
    <li>postgresql12Packages.postgis.doc</li>
    <li>postgresql13JitPackages.postgis</li>
    <li>postgresql13JitPackages.postgis.doc</li>
    <li>postgresql13Packages.postgis</li>
    <li>postgresql13Packages.postgis.doc</li>
    <li>postgresql14JitPackages.postgis</li>
    <li>postgresql14JitPackages.postgis.doc</li>
    <li>postgresql14Packages.postgis</li>
    <li>postgresql14Packages.postgis.doc</li>
    <li>postgresqlJitPackages.postgis (postgresql15JitPackages.postgis)</li>
    <li>postgresqlJitPackages.postgis.doc (postgresql15JitPackages.postgis.doc)</li>
    <li>postgresqlPackages.postgis (postgresql15Packages.postgis)</li>
    <li>postgresqlPackages.postgis.doc (postgresql15Packages.postgis.doc)</li>
    <li>postgresql16JitPackages.postgis</li>
    <li>postgresql16JitPackages.postgis.doc</li>
    <li>postgresql16Packages.postgis</li>
    <li>postgresql16Packages.postgis.doc</li>
    <li>python311Packages.bsuite</li>
    <li>python311Packages.bsuite.dist</li>
    <li>python311Packages.cartopy</li>
    <li>python311Packages.cartopy.dist</li>
    <li>python311Packages.django-bootstrap4</li>
    <li>python311Packages.django-bootstrap4.dist</li>
    <li>python311Packages.fiona</li>
    <li>python311Packages.fiona.dist</li>
    <li>python311Packages.folium</li>
    <li>python311Packages.folium.dist</li>
    <li>python311Packages.geopandas</li>
    <li>python311Packages.geopandas.dist</li>
    <li>python311Packages.plotnine</li>
    <li>python311Packages.plotnine.dist</li>
    <li>python311Packages.pygmt</li>
    <li>python311Packages.pygmt.dist</li>
    <li>python311Packages.python-mapnik</li>
    <li>python311Packages.python-mapnik.dist</li>
    <li>python311Packages.wktutils</li>
    <li>python311Packages.wktutils.dist</li>
    <li>python312Packages.gdal</li>
    <li>python312Packages.pygmt</li>
    <li>python312Packages.pygmt.dist</li>
    <li>pytrainer</li>
    <li>pytrainer.dist</li>
    <li>qgis</li>
    <li>qmapshack</li>
    <li>saga</li>
    <li>sumo</li>
    <li>udig</li>
    <li>vpv</li>
  </ul>
</details>

**Review by imincik — APPROVED** [review]

**Comment by imincik** [comment]
@rollf , thanks for contributing.

**Pull request merged**
